### PR TITLE
MCR-2003 new Derivate draft

### PR DESCRIPTION
**Pull request by sebhofmann**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2003).

### CODE REVIEW

**Review by fluetze — APPROVED** [review]

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on b639d4151b6fe6ef8dbfaf0d38a8987bc9d8dd53! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/4514-pull_request-799

**Comment by rochpaul** [comment]
Tested with a new MIR Installation
mir[mir-master] + mycore[issues/MCR-2003-fix_xsd]


Testing steps

1) Dissertation creation with uploading file

- Login with user in author role
- Create new Dissertation
- Upload file

-----------------------------------

- \mir\data\metadata\mir\mods\0000\00\mir_mods_00000001.xml was created in filesystem
- \mir\data\metadata\mir\derivate\0000\00\mir_derivate_00000001.xml was created in filesystem

-----------------------------------


2) Remove uploaded file (as administrator in standard mir installation)
- Lock on \mir\data\metadata\mir\derivate\0000\00\ directory
- removed derivate dependency exists furthermore in \mir\data\metadata\mir\mods\0000\00\mir_mods_00000001.xml

 <structure>
    <derobjects class="MCRMetaDerivateLinkID">
      <derobject inherited="0" xlink:type="locator" xlink:title="data object from mir_mods_00000001" xlink:label="default" xlink:href="mir_derivate_00000001" mainDoc="Vorlage_2018-11-14.pptx" />
    </derobjects>
  </structure>

- Frontend shows  404 Not Found (See Screenshot MCR-2003-remove-uploaded-file)
MCR-2003-remove-uploaded-file


- Inconsistent state after webapp shutdown
- The derivate is removed but the derivate dependency exists furthermore

**Comment by rochpaul** [comment]
stacktrace for removing derivate mir_derivate_00000002 in document mir_mods_00000002

[stacktrace_test_removement_mir_derivate_00000002.txt](https://github.com/MyCoRe-Org/mycore/files/3116950/stacktrace_test_removement_mir_derivate_00000002.txt)

**Review by kkrebs — CHANGES_REQUESTED** [review]
work in progress, some more testing needed

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRMetadataManager.java [hunk 1]
```diff
@@ -212,9 +211,7 @@ public static void create(final MCRDerivate mcrDerivate) throws MCRPersistenceEx
         fireEvent(mcrDerivate, null, MCREvent.CREATE_EVENT);
 
         // add the link to metadata
-        final MCRMetaDerivateLinkID der = new MCRMetaDerivateLinkID();
-        der.setReference(mcrDerivate.getId().toString(), null, mcrDerivate.getLabel());
-        der.setSubTag("derobject");
+        final MCRMetaDerivateLinkID der = MCRMetaDerivateLinkIDFactory.getInstance().getDerivateLink(mcrDerivate);
```
Comment (kkrebs) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRMetadataManager.java]: this has changed from link to derobject, we need:
-> better documentation
-> better naming

File: mycore-migration/src/main/java/org/mycore/migration/cli/MCRMigrationCommands.java [hunk 2]
```diff
@@ -317,6 +316,39 @@ public static void migrateTEIEntrysOfMetsFileOfDerivate(String derivateIdStr)
         LOGGER.info("Migrated mets of " + derivateIdStr);
     }
 
+    // 2018 -> 2019
+    @MCRCommand(syntax = "migrate MCR-2003 for object {0}")
```
Comment (kkrebs) [inline mycore-migration/src/main/java/org/mycore/migration/cli/MCRMigrationCommands.java]: we should not use ticket number here, should be part of help message
see  help = "Adds missing children to structure of parent {0}. (MCR-1480)",

File: mycore-migration/src/main/java/org/mycore/migration/cli/MCRMigrationCommands.java [hunk 3]
```diff
@@ -317,6 +316,39 @@ public static void migrateTEIEntrysOfMetsFileOfDerivate(String derivateIdStr)
...
+
+        final MCRObject mcrObject = MCRMetadataManager.retrieveMCRObject(objectID);
+        final List<MCRMetaDerivateLinkID> derivates = mcrObject.getStructure().getDerivates();
+
+        return derivates.stream().map(
+            (der) -> "migrate MCR-2003 set order of Derivate " + der.getXLinkHrefID().toString() + " to " + (
+                derivates.indexOf(der) + 1)).collect(Collectors.toList());
+    }
+
+    @MCRCommand(syntax = "migrate MCR-2003 set order of Derivate {0} to {1}")
```
Comment (kkrebs) [inline mycore-migration/src/main/java/org/mycore/migration/cli/MCRMigrationCommands.java]: "Derivate" -> "derivate"

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRDefaultMetaDerivateLinkIDFactory.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.ArrayList;
+
+import org.jdom2.Content;
+import org.jdom2.Element;
+import org.mycore.common.MCRConstants;
+
+public class MCRDefaultMetaDerivateLinkIDFactory extends MCRMetaDerivateLinkIDFactory{
+
+    @Override
+    public MCRMetaDerivateLinkID getDerivateLink(MCRDerivate der) {
```
Comment (kkrebs) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRDefaultMetaDerivateLinkIDFactory.java]: we should think about naming here, e.g.
 - MCRMetaStructureDerObject
 - MCRMetaDerObject

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 8bd944ca1b16b2a55bf669139dada3023551ef41! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/4531-pull_request-799

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 8bd944ca1b16b2a55bf669139dada3023551ef41! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/4531-pull_request-799

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 328844ef6c064c013a274d581a50d4992ea350ed! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/4533-pull_request-799

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 884ae52fe6992d464740cb4f6b26bc14585054c6! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/4541-pull_request-799

**Review by kkrebs — APPROVED** [review]

**Pull request merged**